src/main/java/daomephsta/unpick/impl/constantresolvers/BytecodeAnalysisConstantResolver.java

```diff
@@ -34,36 +55,232 @@ public BytecodeAnalysisConstantResolver(IClassResolver classResolver) {
 
 	@Override
 	public ResolvedConstant resolveConstant(String owner, String name) {
-		return constantDataCache.computeIfAbsent(owner, this::extractConstants).get(name);
+		Map<String, ResolvedConstant> resolvedConstants = constantDataCache.computeIfAbsent(owner, this::extractConstants);
+		return resolvedConstants == null ? null : resolvedConstants.get(name);
 	}
 
-	private ResolvedConstants extractConstants(String owner) {
-		ClassReader cr = classResolver.resolveClass(owner);
-		ResolvedConstants resolvedConstants = new ResolvedConstants(Opcodes.ASM9);
-		cr.accept(resolvedConstants, 0);
-		return resolvedConstants;
+	@Override
+	public Map<String, ResolvedConstant> getAllConstantsInClass(String owner) {
+		Map<String, ResolvedConstant> resolvedConstants = constantDataCache.computeIfAbsent(owner, this::extractConstants);
+		return resolvedConstants == null ? null : Collections.unmodifiableMap(resolvedConstants);
 	}
 
-	private static class ResolvedConstants extends ClassVisitor {
-		ResolvedConstants(int api) {
-			super(api);
+	@Nullable
+	protected Map<String, ResolvedConstant> extractConstants(String owner) {
+		ClassReader cr = classResolver.resolveClass(owner);
+		if (cr == null) {
+			return null;
 		}
+		ResolvedConstantsBuilder builder = new ResolvedConstantsBuilder(owner);
+		cr.accept(builder, 0);
+		return builder.resolvedConstants;
+	}
 
+	// Tries to get the constant value out of the ConstantValue attribute. If the constant is a Class, or null, javac
+	// doesn't seem to store it there, so we fall back to scraping the constants out of the <clinit> or <init> methods.
+	// In both cases, all assignments to the constant must be an instruction representing the same constant value. For
+	// non-static constants, all constructors not delegating via this() must assign the constant.
+	private static class ResolvedConstantsBuilder extends ClassVisitor {
+		private final String ownerClass;
 		private final Map<String, ResolvedConstant> resolvedConstants = new HashMap<>();
+		private final Map<String, Type> staticFieldsNeedingValue = new HashMap<>();
+		private final Map<String, Type> nonStaticFieldsNeedingValue = new HashMap<>();
+		private final List<MethodNode> constructors = new ArrayList<>();
+		@Nullable
+		private MethodNode staticInitializer;
+
+		ResolvedConstantsBuilder(String ownerClass) {
+			super(Opcodes.ASM9);
+			this.ownerClass = ownerClass;
+		}
 
 		@Override
 		public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
-			if (Modifier.isStatic(access) && Modifier.isFinal(access)) {
-				Type fieldType = Type.getType(descriptor);
-				if (VALID_CONSTANT_TYPES.stream().anyMatch(t -> t.equals(fieldType))) {
-					resolvedConstants.put(name, new ResolvedConstant(fieldType, value));
+			if ((access & Opcodes.ACC_FINAL) == 0) {
```

Comment:
Why the shift away from using the `isX` methods of `java.lang.reflect.Modifier`? They're more readable.

Reply:
Because I am paranoid that the flags for reflection modifiers and access flags might at some point start diverging, so I don't want to get in the habit of using `Modifier` to check for access flags, even though it will never break for the existing flags. Maybe I am being overly paranoid though, since the [`Modifier` docs](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/lang/reflect/Modifier.html) do explicitly say they were copied from the access flags.